src/p11_wrap.cpp

```diff
@@ -0,0 +1,274 @@
+{
+    if (slot == nullptr) {
+        throw P11Exception("Cannot open session from NULL slot pointer");
+    }
+
+    P11IsNonNegative::callChecked(
+            lib::LibP11::P11_PKCS11_open_session, slot, rw /* Denotes read/write mode. */);
+}
+
+void _PKCS11_login(PKCS11_SLOT *slot, std::string pin)
+{
+    if (slot == nullptr) {
+        throw P11Exception("Cannot login from NULL slot pointer");
+    }
+
+    if (_PKCS11_is_logged_in(slot)) {
+        throw P11Exception("Mismatch detected; user already logged in");
+    }
+
+    P11IsNonNegative::callChecked(
+            lib::LibP11::P11_PKCS11_login, slot, 0 /* Always login as normal user. */, pin.c_str());
+}
+
+void _PKCS11_logout(PKCS11_SLOT *slot)
+{
+    if (slot == nullptr) {
+        throw P11Exception("Cannot logout from NULL slot pointer");
+    }
+
+    if (!_PKCS11_is_logged_in(slot)) {
+        throw P11Exception("Mismatch detected; user not logged in to perform logout");
+    }
+
+    P11IsNonNegative::callChecked(lib::LibP11::P11_PKCS11_logout, slot);
+}
+
+bool _PKCS11_is_logged_in(PKCS11_SLOT *slot)
+{
+    if (slot == nullptr) {
+        throw P11Exception("Cannot check login from NULL slot pointer");
+    }
+
+    int result;
+    P11IsNonNegative::callChecked(
+            lib::LibP11::P11_PKCS11_is_logged_in, slot, 0 /* Check normal user. */, &result);
+
+    // Result is 1 if logged in.
+    return result == 1;
+}
+
+void _PKCS11_store_private_key(PKCS11_TOKEN *token, EVP_PKEY *pk, std::string label, std::string id)
+{
+    std::string parsedStrID = parseID(id);
+    unsigned char *parsedID =
+            reinterpret_cast<unsigned char *>(const_cast<char *>(parsedStrID.c_str()));
+
+    P11IsNonNegative::callChecked(lib::LibP11::P11_PKCS11_store_private_key,
+                                  token,
+                                  pk,
+                                  const_cast<char *>(label.c_str()),
+                                  parsedID,
+                                  parsedStrID.size());
+}
+
+void _PKCS11_store_public_key(PKCS11_TOKEN *token, EVP_PKEY *pk, std::string label, std::string id)
+{
+    std::string parsedStrID = parseID(id);
+    unsigned char *parsedID =
+            reinterpret_cast<unsigned char *>(const_cast<char *>(parsedStrID.c_str()));
+
+    P11IsNonNegative::callChecked(lib::LibP11::P11_PKCS11_store_public_key,
+                                  token,
+                                  pk,
+                                  const_cast<char *>(label.c_str()),
+                                  parsedID,
+                                  parsedStrID.size());
+}
+
+void _PKCS11_generate_key(PKCS11_TOKEN *token, unsigned int bits, std::string label, std::string id)
+{
+    std::string parsedStrID = parseID(id);
+    unsigned char *parsedID =
+            reinterpret_cast<unsigned char *>(const_cast<char *>(parsedStrID.c_str()));
+
+    // Note, PKCS11_generate_key() function only generates RSA keys.
```

Comment:
Honestly, we don't need to create RSA keys, so I wonder how useful adding a deprecated function is really is for us.

Reply:
From above: 

> ECC is out-of-scope for this PR (it is already big enough). It will be addressed in the next one.